### ISSUE TITLE
Lists are mutable -- aren't they supposed to be immutable?

### ISSUE THREAD

**Issue by krader1961** [elves/elvish#920]
```
~/tmp/elvish> x = [1 2 3]
~/tmp/elvish> put $@x
▶ 1
▶ 2
▶ 3
~/tmp/elvish> x[1] = 99
~/tmp/elvish> put $@x
▶ 1
▶ 99
▶ 3
~/tmp/elvish> del x[1]
Exception: value does not support element removal
[tty], line 1: del x[1]
```

The only mention of this I could find was in https://elv.sh/ref/language.html#list-indexing but not the earlier section that introduces lists. And the `del` documentation implies it works with lists as well as maps.

**Comment by krader1961** [comment]
https://elv.sh/ref/language.html#deleting-variable-or-element-del states explicitly that...

> The del special command can be used to either delete variables or elements of lists and maps.

Which contradicts the text in the "list-indexing" section of that document which states that lists are immutable. Not to mention the current behavior of `del` throws an exception when deleting an element of a list.

This looks like the desired behavior has changed over time in an inconsistent manner. And the documentation is inconsistent because the implementation is inconsistent. So the question is whether Elvish lists should be immutable (like Python tuples) or mutable (like Python lists). I'm inclined to make them mutable; i.e., allow `del` to mutate the list. Yes, that is a potential source of bugs when deleting elements of a list while mutating over it. But the same is true, in theory, for deleting elements of a map. It seems to me you can't have it both ways. That is, why is deleting elements of a map okay during iteration over the map while deleting elements of a list is not?

Yes, I'm aware of the obvious answer that you can iterate over a list in this manner:

```
l = [1 2 3]
for x $l { put $x }
```

Which means that deleting either of the first two elements during the iteration raises the question of what it means to continue iterating over the list. Whereas iterating over the keys of a map requires making an explicit list of the keys:

```
m = [&k1=1 &k2=2]
for x [(keys $m)] { put $m[$x] }
```

The latter is immune to mutating, by deletion, of elements of the map due to the explicit construction of a list containing the keys of the map prior to the modification of the map. Which I can accept but then why can't I `del li[0]` outside of a `for` loop? Yes, that is a somewhat silly rhetorical question. But it illustrates why such questions are difficult to answer when designing a new language.

**Comment by xiaq** [comment]
Elvish's lists and maps have always been immutable, it is the variable binding that is mutable. An element assignment like `li[0] = a` creates a new binding for `li`, and does not mutate the original list. This is explained in https://elv.sh/learn/unique-semantics.html#assignment-semantics, but I should also explain that in the reference doc.

The `del` command does not work with list elements, the doc is wrong to say that it does. There is no efficient way to delete an element from a list as it involves moving all the elements after it.

**Comment by krader1961** [comment]
> An element assignment like li[0] = a creates a new binding for li, and does not mutate the original list.

I understand what you're saying but that is extremely subtle. I would bet that 96% of elvish users are unaware of that behavior and will be surprised by it. I read all the elvish docs a year ago and forgot about the "unique assignment semantics".  Too, the example in https://elv.sh/learn/unique-semantics.html#assignment-semantics, only applies to maps. If you replace the map example with what would seem to be an equivalent example using a list you get the expected nonsensical results:

```
~> l1 = [11 22 33]
~> l2 = $l2
~> put $@l1
▶ 11
▶ 22
▶ 33
~> put $@l2
▶ l
▶ 1
~> l2[1] = 44
~> put $@l2
▶ l
▶ 4
▶ 4
~> put $l2
▶ l44
~> put $l1
▶ [11 22 33]
```

Note that it let me mutate the string assigned to `l2` by extending its length. That, too, seems inconsistent with the idea of immutable data structures.


P.S., That it is inefficient to delete an element from a list does not, by itself, seem like a good reason to disallow it. Especially when you can otherwise mutate the list by essentially binding a new instance of the list to the var via an assignment to an element of the list. It seems to me it would be more consistent to allow `del` but document that it will be expensive and should be avoided.

**Comment by rdw20170120** [comment]
I presume that deleting a list element would be expensive because of the Go data structure (array?) used to implement it.  Should we consider changing that data structure (linked list?) to make `del` reasonably efficient on a list?  I think that `del` would have immense value on lists in many scripts.  It seems a shame to sacrifice that value over a minor performance issue.

**Comment by krader1961** [comment]
I should have closed this long ago. I opened it a few weeks after starting to use Elvish as my primary shell. It reflects my expectations due to my experience with languages like Python and C where arrays/lists are passed by reference. While phrased as a question I was really pointing out what I thought was undesirable behavior; i.e., a design mistake rather than implementation bug. It is a good example why everyone, including me, should avoid opening such issues until they have more than a few weeks of experience with a language. That @xiaq did not close it is because he has far more patience and probably wondered if I would follow up with a better argument for changing the Elvish behavior. Which I obviously never did. 😄 

Note that the documentation regarding this behavior has been improved since I opened this issue. No doubt further improvements would be useful but that should be tackled separate from this issue.

Regarding the performance question asked by @rdw20170120 that will be resolved over time as the focus shifts from stabilizing the language to issues like cross-platform behavior (e.g., improved support for Windows) and performance. There is a lot of prior art regarding immutable, or persistent, data structures. See https://en.wikipedia.org/wiki/Persistent_data_structure for a quick introduction.
